### PR TITLE
chore(rpc): `EthApi` builder

### LINKED ISSUE #8986: Builder pattern for `EthHandlers`
### Describe the feature

Decompose `init_eth` function, using builder pattern

https://github.com/paradigmxyz/reth/blob/e90b526a6668fae51179541221666894b3ade14b/crates/rpc/rpc-builder/src/lib.rs#L1015-L1081

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by emhane**
Closes https://github.com/paradigmxyz/reth/issues/8986

- Adds trait `EthApiBuilder` for building custom RPC server for `eth` namespace
- Rename `RethModuleRegistry` to `RpcRegistryInner`, coherent with other rpc types.
- `EthApiBuilder` only stored as trait object in type in `reth-rpc-builder`, and is constrained by the server type it builds. The server type built is inherited from node builder.
- Removes unused fields from `RpcRegistryInner`. `set_eth_raw_transaction_forwarder` has been removed with precaution: on main, the only way to add the tx forwarder outside of module, is via method `RpcRegistryInner::set_eth_raw_transaction_forwarder`, so it's safe to build `EthHandlers` in `RpcRegistryInner` constructor with field set to None in `EthApiServer` type.

### CODE REVIEW

File: crates/rpc/rpc-builder/src/eth.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/eth.rs:179]: directionally, this makes sense.

although, I think we should try to add another layer of abstraction here

1. we need some type that launches the server (rpc-builder)
2. this will require something that can create an EthApi

so perhaps we limit this one here just to `EthApi` (e.g. EthApiBuilder)

so we can eventually abstract the rpc entirely

**Review by mattsse — CHANGES_REQUESTED** [review]
I like this direction

we should limit this pr to the ethapi abstraction, and in a later step we can integrate all of this in nodebuilder as well.

File: crates/rpc/rpc-builder/src/eth.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:179]: I was pondering on the name and thought `ApiBuilder` sounds like metaprogramming, but I'm fine with changing it

Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:179]: https://en.wikipedia.org/wiki/Metaprogramming

File: crates/node/builder/src/rpc.rs
Comment (tcoratger) [inline crates/node/builder/src/rpc.rs:151]: Maybe we can create type aliases to avoid this allow no?

File: crates/node/builder/src/rpc.rs
Comment (tcoratger) [inline crates/node/builder/src/rpc.rs:7]: Why do we need to change this, this is the same as before no?

File: crates/rpc/rpc-builder/src/eth.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/src/eth.rs:3]:
```suggestion
use std::sync::Arc;
use reth_provider::{BlockReaderIdExt, CanonStateSubscriptions, ChainSpecProvider};
```

File: crates/rpc/rpc-builder/src/eth.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/src/eth.rs:16]:
```suggestion
use reth_tasks::TaskSpawner;
use serde::{Deserialize, Serialize};
use crate::{fee_history_cache_new_blocks_task, RawTransactionForwarder};
```

File: crates/rpc/rpc-builder/src/eth.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/src/eth.rs:171]: As all the fields of the `EthApiBuilderCtx` struct are public, are we sure that we need this kind of initializer with `too_many_arguments`? Maybe we can let the user initialize the structure himself, which amounts to the same thing because here we also have to repeat all the elements one by one, wdyt?

File: crates/rpc/rpc-builder/src/lib.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/src/lib.rs:147]:
```suggestion
use std::{
    collections::HashMap,
    fmt,
    net::{Ipv4Addr, SocketAddr, SocketAddrV4},
    sync::Arc,
    time::{Duration, SystemTime, UNIX_EPOCH},
};
use error::{ConflictingModules, RpcError, ServerKind};
```

File: crates/rpc/rpc-builder/src/lib.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/src/lib.rs:183]:
```suggestion
use tracing::{instrument, trace};
use crate::{
    auth::AuthRpcModule, cors::CorsDomainError, error::WsHttpSamePortError,
    metrics::RpcRequestMetrics,
};
```

File: crates/rpc/rpc-builder/tests/it/utils.rs
Comment (tcoratger) [inline crates/rpc/rpc-builder/tests/it/utils.rs:3]:
```suggestion
use std::net::{Ipv4Addr, SocketAddr, SocketAddrV4};
use reth_beacon_consensus::BeaconConsensusEngineHandle;
```

File: crates/rpc/rpc-eth-api/src/api/servers/mod.rs
Comment (tcoratger) [inline crates/rpc/rpc-eth-api/src/api/servers/mod.rs:6]:
```suggestion
use std::{ops::Deref, sync::Arc};
use derive_more::Deref;
```

File: crates/storage/provider/src/test_utils/noop.rs
Comment (tcoratger) [inline crates/storage/provider/src/test_utils/noop.rs:6]:
```suggestion
use std::{
    ops::{RangeBounds, RangeInclusive},
    sync::Arc,
};
use reth_chainspec::{ChainInfo, ChainSpec, MAINNET};
```

File: crates/storage/provider/src/test_utils/noop.rs
Comment (tcoratger) [inline crates/storage/provider/src/test_utils/noop.rs:27]:
```suggestion
use tokio::sync::broadcast;
use crate::{
    providers::StaticFileProvider,
    traits::{BlockSource, ReceiptProvider},
```

File: crates/node/builder/src/rpc.rs
Comment (emhane) [reply crates/node/builder/src/rpc.rs:7]: useful for readers who care about std/no std

File: crates/rpc/rpc-builder/src/eth.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:3]: if whitespace is removed then fmt orders alphabetically

File: crates/rpc/rpc-builder/src/eth.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:171]: this pr is one part in a big change to rpc design, so it is still prone to change. that's why it make sense to not spend time on making it perfect just yet.

File: crates/rpc/rpc-builder/src/eth.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:3]: yes! I love this, very helpful. we want [StdExternalCrate](https://rust-lang.github.io/rustfmt/?version=v1.6.0&search=#StdExternalCrate%5C%3A). feel free to open an issue.

File: crates/node/builder/src/rpc.rs
Comment (emhane) [reply crates/node/builder/src/rpc.rs:151]: ref https://github.com/paradigmxyz/reth/pull/9041#discussion_r1654457204

**Comment by emhane** [comment]
re-request review @mattsse

File: crates/ethereum/rpc/src/api_builder.rs
Comment (mattsse) [inline crates/ethereum/rpc/src/api_builder.rs:29]: I think we should shift this to FullNode trait and not use individual generics, given that we want to bake this into node builder anyway

**Review by mattsse — CHANGES_REQUESTED** [review]
I like the direction, this is basically a builder trait for the EthApi.

I'd like to make this consistent with the other builder traits we already use

https://github.com/paradigmxyz/reth/blob/6d8cbae3e6d4b441b04af06bbbdba0511993d036/crates/node/builder/src/components/pool.rs#L7-L7

but not generic over types and instead:

https://github.com/paradigmxyz/reth/blob/6d8cbae3e6d4b441b04af06bbbdba0511993d036/crates/node/builder/src/rpc.rs#L33-L33

File: crates/ethereum/rpc/src/api_builder.rs
Comment (emhane) [reply crates/ethereum/rpc/src/api_builder.rs:29]: the intention is to implement the builder pattern in a next pr, which will use the `FullNodeComponents`, the problem here otherwise is cyclical dependencies between `reth-rpc-builder` and `reth-node-builder`.

I can move the implementation of trait `reth_rpc_builder::EthApiBuilder` into `reth-rpc-builder`, however, it is Layer one specific. so I'm wondering if eventually it would make more sense in the new `reth-ethereum-rpc` crate.

File: crates/ethereum/rpc/src/api_builder.rs
Comment (mattsse) [inline crates/ethereum/rpc/src/api_builder.rs:13]:
```suggestion
#[non_exhaustive]
pub struct EthApiBuild;
```

File: crates/ethereum/rpc/src/api_builder.rs
Comment (mattsse) [inline crates/ethereum/rpc/src/api_builder.rs:25]:
```suggestion
    type EthApi = EthApi<Provider, Pool, Network, EvmConfig>;
```

File: crates/ethereum/rpc/src/api_builder.rs
Comment (mattsse) [inline crates/ethereum/rpc/src/api_builder.rs:13]: we could name this EthEthApiBuilder :D but that would be silly -.-

File: crates/rpc/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:428]: can we use `B` and put this in a where clause?

File: crates/rpc/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:489]: same

File: crates/rpc/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:501]: can we rename this to `build_with` and keep the existing `build` fn that calls `build_with` with the EthApiBuild?

**Review by mattsse — CHANGES_REQUESTED** [review]
gotcha, then this is the next intermediary step towards nodebuilder integration and we need need to take a look at how to separate rpc-builder from node-builder, or at least separate the node-builder internal traits

File: crates/ethereum/rpc/src/api_builder.rs
Comment (emhane) [reply crates/ethereum/rpc/src/api_builder.rs:13]: haha yeah that's why I like to name traits with imperative verbs like `BuildEthApi`, so the noun I reserved for the type

**Comment by emhane** [comment]
I simplified this based on latest review and general goal of configuring relevant rpc modules in node builder.

next methods to work on in `reth-rpc-builder` scope are  `RethRegistry::reth_methods`, `RethRegistry::module_for` and `RethRegistry::maybe_module`. rn they are the funnel through which it is required for the configured `EthApi` type to impl `FullEthApiServer` - which is a shame for users who just want to enable the `AdminApiServer` for example, or `EthApiServer` but without calling tracing methods.

I think it makes sense to continue in the node builder scope next. let's get the code in place required to add certain modules, even though rn it will be required that the configured `EthApi` type supports adding all modules (even if user just wants to add one little module).

File: crates/rpc/rpc-builder/src/eth.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/eth.rs:307]: why do we need these additional builder types now?

File: crates/rpc/rpc-builder/src/eth.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/eth.rs:180]: did we replace these functions with builder structs?
why?

File: crates/rpc/rpc-eth-api/src/helpers/mod.rs
Comment (mattsse) [inline crates/rpc/rpc-eth-api/src/helpers/mod.rs:50]: can we drop the `ServerGateway` here?

File: crates/rpc/rpc-builder/src/lib.rs
Comment (mattsse) [inline crates/rpc/rpc-builder/src/lib.rs:511]: can we add `build_ethereum` that calls this with the ethereum builder?

**Review by mattsse — CHANGES_REQUESTED** [review]
last nits, questions.

this should unblock us from integrating this in node-builder and creating an op impl

File: crates/rpc/rpc-builder/src/eth.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/eth.rs:180]: in anticipation that we will make those apis configurable in the future too, exposing all eth handlers to user (and I had already implemented it before Thomas pr was merged)

File: crates/rpc/rpc-eth-api/src/helpers/mod.rs
Comment (emhane) [reply crates/rpc/rpc-eth-api/src/helpers/mod.rs:50]: > An API gateway is a data-plane entry point for API calls that represent client requests to target applications and services.

the descriptive name is very helpful. would you be ok with `FullEthApiGateway`?

Comment (emhane) [reply crates/rpc/rpc-eth-api/src/helpers/mod.rs:50]: https://www.f5.com/glossary/api-gateway

File: crates/rpc/rpc-builder/src/lib.rs
Comment (emhane) [reply crates/rpc/rpc-builder/src/lib.rs:511]: not yet I think, since rpc isn't passed down from that level yet. or wdym?

**Review by mattsse — APPROVED** [review]
last naming nit

**Pull request merged**
